lib/Basics/expected.h

```diff
@@ -0,0 +1,215 @@
+#pragma once
+#include <stdexcept>
+#include <type_traits>
+#include "Assertions/Assert.h"
+
+namespace arangodb {
+template<typename T>
+struct expected {
+  expected() noexcept {}
+
+  template<typename... Args>
+  explicit expected(std::in_place_t, Args&&... args) noexcept(
+      std::is_nothrow_constructible_v<T, Args...>) requires
+      std::constructible_from<T, Args...> {
+    new (&_value) T(std::forward<Args>(args)...);
+    _state = kValue;
+  }
+
+  explicit expected(std::exception_ptr ex) noexcept
+      : _exception(std::move(ex)), _state(kException) {
+    static_assert(std::is_nothrow_move_constructible_v<std::exception_ptr>);
+  }
+
+  expected(expected const& other) noexcept(
+      std::is_nothrow_copy_constructible_v<T>) requires
+      std::is_copy_constructible_v<T> {
+    if (other._state == kValue) {
+      new (&_value) T(other._value);
+      _state = kValue;
+    } else if (other._state == kException) {
+      static_assert(std::is_nothrow_copy_constructible_v<std::exception_ptr>);
```

Comment:
It would be sufficient to move this out of the constructor and into the class and do it just once.

Reply:
Can you specify "this" more?

Reply:
Ah, I actually referred to just the last line, i.e., the static_assert. The same static_assert is used in several places, but since it is static it would be sufficient to move it into the class body and do it just once.

Reply:
We finally decided to leave the static_asserts where they are because this way it is directly visible why they are needed.